### ISSUE TITLE
Investigate how batching in `merge` affects performance

### ISSUE THREAD

**Issue by inetic** [equalitie/ouisync#143]
I noticed that when syncing a large file (e.g. 30MB) among three nodes, the `merge` algorithm takes a long time to finish. I played with batching inside [`src/lib/blob/mod.rs/Blob::fork`](https://github.com/equalitie/ouisync/blob/b33375d2b2d92e11936a5b66a0c0d9f182b96784/lib/src/blob/mod.rs#L490-L581), but it seems that while it improves the performance of the [`sync_three_peers_one_repo_huge` test](https://github.com/equalitie/ouisync/blob/b33375d2b2d92e11936a5b66a0c0d9f182b96784/lib/tests/sync.rs#L54-L56) it decreases the performance of the [`sync_two_peers_one_repo_huge` test](https://github.com/equalitie/ouisync/blob/b33375d2b2d92e11936a5b66a0c0d9f182b96784/lib/tests/sync.rs#L44-L46). 

Attached are measurements, twice for every case to account for flukes.
[ouisync-stats.txt](https://github.com/equalitie/ouisync/files/12868752/ouisync-stats.txt)

**Comment by inetic** [comment]
It turned out that the two node test did not finish for the long fork to finish and thus it wasn't accounted for in the final stats. I modified the test to make sure fork is done [here](https://github.com/equalitie/ouisync/commit/ccf093acaa990c3250edd63a21701205b95cab4e) and the two node test also shows having benefit from the batching inside `Blob::fork`:

`sync_two_peers_one_repo_huge` with batch size 1
```
┌────────────────────────┬───────┬────────┬──────────┬─────────┬─────────┬────────┬─────────┬──────────┬──────────┐
│ name                   ┆ count ┆    min ┆      max ┆    mean ┆   stdev ┆    50% ┆     90% ┆      99% ┆    99.9% │
╞════════════════════════╪═══════╪════════╪══════════╪═════════╪═════════╪════════╪═════════╪══════════╪══════════╡
│ handle_response        ┆  3302 ┆ 0.0012 ┆   1.6190 ┆  0.0377 ┆  0.0642 ┆ 0.0097 ┆  0.1374 ┆   0.1940 ┆   0.4740 │
│ handle_root_node       ┆   126 ┆ 0.0012 ┆   0.1243 ┆  0.0099 ┆  0.0141 ┆ 0.0083 ┆  0.0161 ┆   0.0771 ┆   0.1243 │
│ handle_inner_node      ┆  1239 ┆ 0.0029 ┆   1.6190 ┆  0.0143 ┆  0.0639 ┆ 0.0084 ┆  0.0158 ┆   0.0818 ┆   0.9899 │
│ handle_leaf_node       ┆   969 ┆ 0.0037 ┆   0.0474 ┆  0.0069 ┆  0.0033 ┆ 0.0062 ┆  0.0098 ┆   0.0203 ┆   0.0474 │
│ handle_block           ┆   968 ┆ 0.0047 ┆   0.2653 ┆  0.1005 ┆  0.0557 ┆ 0.1043 ┆  0.1709 ┆   0.2003 ┆   0.2653 │
│ handle_block_not_found ┆     0 ┆ 0.0000 ┆   0.0000 ┆  0.0000 ┆  0.0000 ┆ 0.0000 ┆  0.0000 ┆   0.0000 ┆   0.0000 │
│ request queued         ┆  3184 ┆ 0.0001 ┆  81.0675 ┆ 12.5726 ┆ 16.3017 ┆ 7.6169 ┆ 34.3597 ┆  75.6988 ┆  80.5306 │
│ request inflight       ┆  3184 ┆ 0.0112 ┆   0.2129 ┆  0.0511 ┆  0.0270 ┆ 0.0464 ┆  0.0860 ┆   0.1489 ┆   0.2118 │
│ handle_request         ┆  3184 ┆ 0.0002 ┆   0.0477 ┆  0.0023 ┆  0.0043 ┆ 0.0012 ┆  0.0030 ┆   0.0214 ┆   0.0446 │
│ scan                   ┆    52 ┆ 0.0021 ┆   7.7846 ┆  2.4979 ┆  1.6999 ┆ 3.2044 ┆  3.8923 ┆   7.7846 ┆   7.7846 │
│ merge                  ┆    27 ┆ 0.0020 ┆ 102.5423 ┆  3.8156 ┆ 19.3093 ┆ 0.0273 ┆  0.0676 ┆ 102.5423 ┆ 102.5423 │
│ prune                  ┆    21 ┆ 0.0008 ┆   0.0663 ┆  0.0139 ┆  0.0184 ┆ 0.0053 ┆  0.0246 ┆   0.0663 ┆   0.0663 │
│ trash                  ┆    21 ┆ 0.0012 ┆   6.4425 ┆  1.3813 ┆  1.5797 ┆ 0.8221 ┆  2.8186 ┆   6.4425 ┆   6.4425 │
└────────────────────────┴───────┴────────┴──────────┴─────────┴─────────┴────────┴─────────┴──────────┴──────────┘```

`sync_two_peers_one_repo_huge` with batch size 32
┌────────────────────────┬───────┬────────┬─────────┬────────┬─────────┬────────┬─────────┬─────────┬─────────┐
│ name                   ┆ count ┆    min ┆     max ┆   mean ┆   stdev ┆    50% ┆     90% ┆     99% ┆   99.9% │
╞════════════════════════╪═══════╪════════╪═════════╪════════╪═════════╪════════╪═════════╪═════════╪═════════╡
│ handle_response        ┆  4059 ┆ 0.0010 ┆  1.1828 ┆ 0.0199 ┆  0.0664 ┆ 0.0093 ┆  0.0261 ┆  0.3125 ┆  0.8766 │
│ handle_root_node       ┆   101 ┆ 0.0010 ┆  0.0692 ┆ 0.0080 ┆  0.0108 ┆ 0.0032 ┆  0.0253 ┆  0.0377 ┆  0.0692 │
│ handle_inner_node      ┆  2015 ┆ 0.0023 ┆  0.9060 ┆ 0.0143 ┆  0.0467 ┆ 0.0081 ┆  0.0165 ┆  0.1856 ┆  0.8431 │
│ handle_leaf_node       ┆   972 ┆ 0.0038 ┆  0.0634 ┆ 0.0071 ┆  0.0043 ┆ 0.0061 ┆  0.0104 ┆  0.0291 ┆  0.0634 │
│ handle_block           ┆   970 ┆ 0.0053 ┆  1.1828 ┆ 0.0428 ┆  0.1144 ┆ 0.0147 ┆  0.0608 ┆  0.7508 ┆  1.1828 │
│ handle_block_not_found ┆     1 ┆ 0.0023 ┆  0.0023 ┆ 0.0023 ┆  0.0000 ┆ 0.0023 ┆  0.0023 ┆  0.0023 ┆  0.0023 │
│ request queued         ┆  4083 ┆ 0.0000 ┆ 37.5810 ┆ 9.4367 ┆ 11.1182 ┆ 6.7444 ┆ 30.7359 ┆ 37.0441 ┆ 37.5810 │
│ request inflight       ┆  4082 ┆ 0.0058 ┆  3.9762 ┆ 0.0606 ┆  0.0716 ┆ 0.0522 ┆  0.0975 ┆  0.2538 ┆  0.3230 │
│ handle_request         ┆  4082 ┆ 0.0002 ┆  0.0412 ┆ 0.0043 ┆  0.0057 ┆ 0.0018 ┆  0.0123 ┆  0.0286 ┆  0.0372 │
│ scan                   ┆    44 ┆ 0.0056 ┆  5.9391 ┆ 1.7010 ┆  1.7414 ┆ 0.7172 ┆  4.3285 ┆  5.9391 ┆  5.9391 │
│ merge                  ┆    46 ┆ 0.0016 ┆ 20.8037 ┆ 0.4864 ┆  3.0191 ┆ 0.0233 ┆  0.0655 ┆ 20.8037 ┆ 20.8037 │
│ prune                  ┆    44 ┆ 0.0002 ┆  0.5222 ┆ 0.0253 ┆  0.0770 ┆ 0.0072 ┆  0.0385 ┆  0.5222 ┆  0.5222 │
│ trash                  ┆    36 ┆ 0.0012 ┆  4.5298 ┆ 1.3160 ┆  1.2365 ┆ 0.7927 ┆  2.7179 ┆  4.5298 ┆  4.5298 │
└────────────────────────┴───────┴────────┴─────────┴────────┴─────────┴────────┴─────────┴─────────┴─────────┘
```
`sync_two_peers_one_repo_huge` with batch size 512
```
┌────────────────────────┬───────┬────────┬─────────┬────────┬────────┬────────┬─────────┬─────────┬─────────┐
│ name                   ┆ count ┆    min ┆     max ┆   mean ┆  stdev ┆    50% ┆     90% ┆     99% ┆   99.9% │
╞════════════════════════╪═══════╪════════╪═════════╪════════╪════════╪════════╪═════════╪═════════╪═════════╡
│ handle_response        ┆  3406 ┆ 0.0013 ┆  5.4358 ┆ 0.0168 ┆ 0.1221 ┆ 0.0090 ┆  0.0214 ┆  0.0771 ┆  0.6333 │
│ handle_root_node       ┆    49 ┆ 0.0011 ┆  0.0771 ┆ 0.0126 ┆ 0.0139 ┆ 0.0078 ┆  0.0258 ┆  0.0771 ┆  0.0771 │
│ handle_inner_node      ┆  1420 ┆ 0.0025 ┆  0.6333 ┆ 0.0115 ┆ 0.0343 ┆ 0.0075 ┆  0.0137 ┆  0.1049 ┆  0.6291 │
│ handle_leaf_node       ┆   969 ┆ 0.0030 ┆  1.9545 ┆ 0.0093 ┆ 0.0628 ┆ 0.0063 ┆  0.0106 ┆  0.0266 ┆  1.9545 │
│ handle_block           ┆   967 ┆ 0.0048 ┆  5.4358 ┆ 0.0306 ┆ 0.2157 ┆ 0.0123 ┆  0.0527 ┆  0.0792 ┆  5.4358 │
│ handle_block_not_found ┆     1 ┆ 0.0022 ┆  0.0022 ┆ 0.0022 ┆ 0.0000 ┆ 0.0022 ┆  0.0022 ┆  0.0022 ┆  0.0022 │
│ request queued         ┆  3365 ┆ 0.0001 ┆ 26.4409 ┆ 8.4679 ┆ 6.7264 ┆ 6.6102 ┆ 20.2669 ┆ 25.7698 ┆ 26.3067 │
│ request inflight       ┆  3365 ┆ 0.0085 ┆  3.9762 ┆ 0.0569 ┆ 0.0786 ┆ 0.0472 ┆  0.0886 ┆  0.3104 ┆  0.3649 │
│ handle_request         ┆  3365 ┆ 0.0002 ┆  0.0823 ┆ 0.0047 ┆ 0.0085 ┆ 0.0016 ┆  0.0121 ┆  0.0474 ┆  0.0797 │
│ scan                   ┆    30 ┆ 0.0030 ┆  5.5365 ┆ 1.5428 ┆ 1.6034 ┆ 1.0654 ┆  3.3890 ┆  5.5365 ┆  5.5365 │
│ merge                  ┆    32 ┆ 0.0006 ┆ 10.3348 ┆ 0.3451 ┆ 1.7883 ┆ 0.0216 ┆  0.0488 ┆ 10.3348 ┆ 10.3348 │
│ prune                  ┆    27 ┆ 0.0006 ┆  0.4131 ┆ 0.0284 ┆ 0.0771 ┆ 0.0078 ┆  0.0535 ┆  0.4131 ┆  0.4131 │
│ trash                  ┆    26 ┆ 0.0011 ┆  3.1373 ┆ 1.3066 ┆ 1.1750 ┆ 0.9521 ┆  2.7179 ┆  3.1373 ┆  3.1373 │
└────────────────────────┴───────┴────────┴─────────┴────────┴────────┴────────┴─────────┴─────────┴─────────┘
```
`sync_two_peers_one_repo_huge` with batch size 2048
```
┌────────────────────────┬───────┬────────┬─────────┬────────┬────────┬────────┬─────────┬─────────┬─────────┐
│ name                   ┆ count ┆    min ┆     max ┆   mean ┆  stdev ┆    50% ┆     90% ┆     99% ┆   99.9% │
╞════════════════════════╪═══════╪════════╪═════════╪════════╪════════╪════════╪═════════╪═════════╪═════════╡
│ handle_response        ┆  3223 ┆ 0.0015 ┆  7.0800 ┆ 0.0167 ┆ 0.1262 ┆ 0.0102 ┆  0.0266 ┆  0.0543 ┆  0.3355 │
│ handle_root_node       ┆    50 ┆ 0.0014 ┆  0.0186 ┆ 0.0055 ┆ 0.0048 ┆ 0.0027 ┆  0.0136 ┆  0.0186 ┆  0.0186 │
│ handle_inner_node      ┆  1236 ┆ 0.0030 ┆  0.8137 ┆ 0.0124 ┆ 0.0330 ┆ 0.0090 ┆  0.0145 ┆  0.0897 ┆  0.6040 │
│ handle_leaf_node       ┆   969 ┆ 0.0040 ┆  0.0923 ┆ 0.0084 ┆ 0.0047 ┆ 0.0079 ┆  0.0106 ┆  0.0227 ┆  0.0923 │
│ handle_block           ┆   968 ┆ 0.0055 ┆  7.0800 ┆ 0.0293 ┆ 0.2266 ┆ 0.0211 ┆  0.0343 ┆  0.0760 ┆  7.0800 │
│ handle_block_not_found ┆     0 ┆ 0.0000 ┆  0.0000 ┆ 0.0000 ┆ 0.0000 ┆ 0.0000 ┆  0.0000 ┆  0.0000 ┆  0.0000 │
│ request queued         ┆  3181 ┆ 0.0001 ┆ 20.6695 ┆ 7.8325 ┆ 5.0848 ┆ 7.5833 ┆ 16.7101 ┆ 20.2669 ┆ 20.6695 │
│ request inflight       ┆  3181 ┆ 0.0026 ┆  0.2401 ┆ 0.0181 ┆ 0.0274 ┆ 0.0083 ┆  0.0359 ┆  0.2097 ┆  0.2349 │
│ handle_request         ┆  3181 ┆ 0.0003 ┆  0.0307 ┆ 0.0011 ┆ 0.0024 ┆ 0.0007 ┆  0.0011 ┆  0.0174 ┆  0.0269 │
│ scan                   ┆    31 ┆ 0.0010 ┆  7.8182 ┆ 1.5474 ┆ 1.9398 ┆ 0.5243 ┆  3.3890 ┆  7.8182 ┆  7.8182 │
│ merge                  ┆    35 ┆ 0.0026 ┆  7.4491 ┆ 0.2377 ┆ 1.2341 ┆ 0.0164 ┆  0.0692 ┆  7.4491 ┆  7.4491 │
│ prune                  ┆    33 ┆ 0.0002 ┆  0.1772 ┆ 0.0200 ┆ 0.0438 ┆ 0.0045 ┆  0.0257 ┆  0.1772 ┆  0.1772 │
│ trash                  ┆    29 ┆ 0.0007 ┆  6.2076 ┆ 1.2897 ┆ 1.4966 ┆ 0.7801 ┆  2.9696 ┆  6.2076 ┆  6.2076 │
└────────────────────────┴───────┴────────┴─────────┴────────┴────────┴────────┴─────────┴─────────┴─────────┘
```

**Comment by inetic** [comment]
After further adding a [small fix](https://github.com/equalitie/ouisync/commit/f96344daba4d54ea02d2bfcdad48899ff69339ac) to the test, here are the results of the three node test with varying batch sizes in `Blob::fork`.

`sync_three_peers_one_repo_huge` with batch size 1
```
┌────────────────────────┬───────┬────────┬──────────┬─────────┬─────────┬────────┬─────────┬──────────┬──────────┐
│ name                   ┆ count ┆    min ┆      max ┆    mean ┆   stdev ┆    50% ┆     90% ┆      99% ┆    99.9% │
╞════════════════════════╪═══════╪════════╪══════════╪═════════╪═════════╪════════╪═════════╪══════════╪══════════╡
│ handle_response        ┆  8349 ┆ 0.0011 ┆   1.1912 ┆  0.0864 ┆  0.1247 ┆ 0.0160 ┆  0.2611 ┆   0.5075 ┆   0.7969 │
│ handle_root_node       ┆   837 ┆ 0.0011 ┆   0.6921 ┆  0.0715 ┆  0.0952 ┆ 0.0145 ┆  0.2254 ┆   0.2957 ┆   0.6921 │
│ handle_inner_node      ┆  3618 ┆ 0.0023 ┆   1.1912 ┆  0.0786 ┆  0.1312 ┆ 0.0144 ┆  0.2705 ┆   0.5369 ┆   0.8263 │
│ handle_leaf_node       ┆  1950 ┆ 0.0030 ┆   0.3418 ┆  0.0101 ┆  0.0090 ┆ 0.0091 ┆  0.0143 ┆   0.0197 ┆   0.1510 │
│ handle_block           ┆  1944 ┆ 0.0042 ┆   0.7508 ┆  0.1800 ┆  0.1259 ┆ 0.1657 ┆  0.3712 ┆   0.5201 ┆   0.5746 │
│ handle_block_not_found ┆     0 ┆ 0.0000 ┆   0.0000 ┆  0.0000 ┆  0.0000 ┆ 0.0000 ┆  0.0000 ┆   0.0000 ┆   0.0000 │
│ request queued         ┆  7545 ┆ 0.0000 ┆ 153.5451 ┆ 15.9985 ┆ 27.8102 ┆ 9.2610 ┆ 48.3184 ┆ 132.0702 ┆ 151.3976 │
│ request inflight       ┆  7543 ┆ 0.0052 ┆   0.3293 ┆  0.0585 ┆  0.0319 ┆ 0.0543 ┆  0.0928 ┆   0.1887 ┆   0.3146 │
│ handle_request         ┆  7543 ┆ 0.0002 ┆   0.0595 ┆  0.0025 ┆  0.0037 ┆ 0.0015 ┆  0.0045 ┆   0.0198 ┆   0.0451 │
│ scan                   ┆   122 ┆ 0.0007 ┆  12.6836 ┆  3.6554 ┆  2.6338 ┆ 4.9325 ┆  5.7378 ┆  11.8112 ┆  12.6836 │
│ merge                  ┆    65 ┆ 0.0008 ┆ 184.6836 ┆  5.8879 ┆ 30.9240 ┆ 0.0488 ┆  0.1290 ┆ 184.6836 ┆ 184.6836 │
│ prune                  ┆    54 ┆ 0.0002 ┆   4.8654 ┆  0.1441 ┆  0.6766 ┆ 0.0088 ┆  0.0886 ┆   4.8654 ┆   4.8654 │
│ trash                  ┆    46 ┆ 0.0008 ┆  10.7374 ┆  1.8252 ┆  2.4181 ┆ 1.1073 ┆  4.2446 ┆  10.7374 ┆  10.7374 │
└────────────────────────┴───────┴────────┴──────────┴─────────┴─────────┴────────┴─────────┴──────────┴──────────┘

```
`sync_three_peers_one_repo_huge` with batch size 32
```
┌────────────────────────┬───────┬────────┬─────────┬─────────┬─────────┬────────┬─────────┬─────────┬─────────┐
│ name                   ┆ count ┆    min ┆     max ┆    mean ┆   stdev ┆    50% ┆     90% ┆     99% ┆   99.9% │
╞════════════════════════╪═══════╪════════╪═════════╪═════════╪═════════╪════════╪═════════╪═════════╪═════════╡
│ handle_response        ┆  9223 ┆ 0.0011 ┆  2.1391 ┆  0.0452 ┆  0.1232 ┆ 0.0182 ┆  0.0661 ┆  0.7172 ┆  1.3422 │
│ handle_root_node       ┆   550 ┆ 0.0011 ┆  1.4596 ┆  0.0684 ┆  0.1948 ┆ 0.0157 ┆  0.0655 ┆  1.0737 ┆  1.4596 │
│ handle_inner_node      ┆  4691 ┆ 0.0025 ┆  2.1391 ┆  0.0425 ┆  0.1153 ┆ 0.0172 ┆  0.0608 ┆  0.6291 ┆  1.2918 │
│ handle_leaf_node       ┆  2030 ┆ 0.0028 ┆  0.7801 ┆  0.0116 ┆  0.0220 ┆ 0.0096 ┆  0.0157 ┆  0.0375 ┆  0.1085 │
│ handle_block           ┆  1952 ┆ 0.0043 ┆  2.1139 ┆  0.0741 ┆  0.1617 ┆ 0.0380 ┆  0.0834 ┆  1.0821 ┆  1.4596 │
│ handle_block_not_found ┆     0 ┆ 0.0000 ┆  0.0000 ┆  0.0000 ┆  0.0000 ┆ 0.0000 ┆  0.0000 ┆  0.0000 ┆  0.0000 │
│ request queued         ┆  8860 ┆ 0.0000 ┆ 73.0144 ┆ 10.7160 ┆ 14.9506 ┆ 7.7846 ┆ 30.3332 ┆ 70.3301 ┆ 73.0144 │
│ request inflight       ┆  8849 ┆ 0.0091 ┆  0.3775 ┆  0.0695 ┆  0.0370 ┆ 0.0627 ┆  0.1148 ┆  0.1877 ┆  0.3460 │
│ handle_request         ┆  8849 ┆ 0.0002 ┆  0.0354 ┆  0.0032 ┆  0.0036 ┆ 0.0019 ┆  0.0088 ┆  0.0157 ┆  0.0274 │
│ scan                   ┆    90 ┆ 0.0007 ┆ 17.9852 ┆  3.0430 ┆  4.0678 ┆ 1.1996 ┆  8.5899 ┆ 17.9852 ┆ 17.9852 │
│ merge                  ┆   320 ┆ 0.0007 ┆ 36.2388 ┆  0.2481 ┆  2.1061 ┆ 0.0469 ┆  0.1667 ┆  3.6239 ┆ 36.2388 │
│ prune                  ┆   303 ┆ 0.0002 ┆  5.2009 ┆  0.0723 ┆  0.3515 ┆ 0.0020 ┆  0.1384 ┆  1.2583 ┆  5.2009 │
│ trash                  ┆   292 ┆ 0.0010 ┆ 10.8716 ┆  0.7535 ┆  1.3454 ┆ 0.2370 ┆  1.6609 ┆  7.4826 ┆ 10.8716 │
└────────────────────────┴───────┴────────┴─────────┴─────────┴─────────┴────────┴─────────┴─────────┴─────────┘
```
`sync_three_peers_one_repo_huge` with batch size 512
```
┌────────────────────────┬───────┬────────┬─────────┬─────────┬─────────┬────────┬─────────┬─────────┬─────────┐
│ name                   ┆ count ┆    min ┆     max ┆    mean ┆   stdev ┆    50% ┆     90% ┆     99% ┆   99.9% │
╞════════════════════════╪═══════╪════════╪═════════╪═════════╪═════════╪════════╪═════════╪═════════╪═════════╡
│ handle_response        ┆  8493 ┆ 0.0013 ┆  7.9524 ┆  0.0646 ┆  0.2433 ┆ 0.0159 ┆  0.0907 ┆  0.5453 ┆  1.6190 │
│ handle_root_node       ┆   899 ┆ 0.0013 ┆  7.9524 ┆  0.0485 ┆  0.3433 ┆ 0.0138 ┆  0.0459 ┆  0.5495 ┆  7.9524 │
│ handle_inner_node      ┆  3693 ┆ 0.0023 ┆  7.3484 ┆  0.0666 ┆  0.1835 ┆ 0.0144 ┆  0.2999 ┆  0.5620 ┆  1.3673 │
│ handle_leaf_node       ┆  1953 ┆ 0.0027 ┆  0.5704 ┆  0.0101 ┆  0.0141 ┆ 0.0088 ┆  0.0140 ┆  0.0211 ┆  0.2118 │
│ handle_block           ┆  1948 ┆ 0.0043 ┆  7.9188 ┆  0.1181 ┆  0.3653 ┆ 0.0438 ┆  0.3649 ┆  0.5495 ┆  7.3820 │
│ handle_block_not_found ┆     0 ┆ 0.0000 ┆  0.0000 ┆  0.0000 ┆  0.0000 ┆ 0.0000 ┆  0.0000 ┆  0.0000 ┆  0.0000 │
│ request queued         ┆  7635 ┆ 0.0000 ┆ 95.5630 ┆ 15.3319 ┆ 23.0532 ┆ 9.1939 ┆ 59.3242 ┆ 91.2681 ┆ 95.5630 │
│ request inflight       ┆  7629 ┆ 0.0130 ┆  0.3439 ┆  0.0634 ┆  0.0359 ┆ 0.0569 ┆  0.1043 ┆  0.2129 ┆  0.3355 │
│ handle_request         ┆  7629 ┆ 0.0002 ┆  0.0744 ┆  0.0033 ┆  0.0050 ┆ 0.0018 ┆  0.0087 ┆  0.0184 ┆  0.0647 │
│ scan                   ┆    89 ┆ 0.0008 ┆ 14.0258 ┆  3.4099 ┆  3.9456 ┆ 1.9462 ┆  9.3952 ┆ 14.0258 ┆ 14.0258 │
│ merge                  ┆   110 ┆ 0.0030 ┆ 14.8311 ┆  0.3219 ┆  1.9140 ┆ 0.0435 ┆  0.1064 ┆ 13.9586 ┆ 14.8311 │
│ prune                  ┆    98 ┆ 0.0002 ┆  2.1391 ┆  0.1872 ┆  0.3636 ┆ 0.0144 ┆  0.6082 ┆  2.1391 ┆  2.1391 │
│ trash                  ┆    90 ┆ 0.0008 ┆ 10.8045 ┆  2.6471 ┆  3.2109 ┆ 0.3922 ┆  7.3149 ┆ 10.8045 ┆ 10.8045 │
└────────────────────────┴───────┴────────┴─────────┴─────────┴─────────┴────────┴─────────┴─────────┴─────────┘
```

`sync_three_peers_one_repo_huge` with batch size 2048
```
┌────────────────────────┬───────┬────────┬─────────┬─────────┬─────────┬────────┬─────────┬─────────┬─────────┐
│ name                   ┆ count ┆    min ┆     max ┆    mean ┆   stdev ┆    50% ┆     90% ┆     99% ┆   99.9% │
╞════════════════════════╪═══════╪════════╪═════════╪═════════╪═════════╪════════╪═════════╪═════════╪═════════╡
│ handle_response        ┆  8013 ┆ 0.0012 ┆ 13.8915 ┆  0.0531 ┆  0.2869 ┆ 0.0172 ┆  0.0797 ┆  0.5704 ┆  1.0737 │
│ handle_root_node       ┆   563 ┆ 0.0011 ┆  9.9321 ┆  0.0621 ┆  0.4302 ┆ 0.0193 ┆  0.0545 ┆  0.7088 ┆  9.9321 │
│ handle_inner_node      ┆  3549 ┆ 0.0019 ┆ 13.8915 ┆  0.0552 ┆  0.2613 ┆ 0.0153 ┆  0.0818 ┆  0.6291 ┆  1.3757 │
│ handle_leaf_node       ┆  1957 ┆ 0.0033 ┆  0.7214 ┆  0.0108 ┆  0.0171 ┆ 0.0091 ┆  0.0147 ┆  0.0367 ┆  0.0876 │
│ handle_block           ┆  1944 ┆ 0.0037 ┆ 13.8915 ┆  0.0840 ┆  0.3973 ┆ 0.0461 ┆  0.0944 ┆  0.5620 ┆  9.9321 │
│ handle_block_not_found ┆     0 ┆ 0.0000 ┆  0.0000 ┆  0.0000 ┆  0.0000 ┆ 0.0000 ┆  0.0000 ┆  0.0000 ┆  0.0000 │
│ request queued         ┆  7552 ┆ 0.0000 ┆ 65.4983 ┆ 13.1599 ┆ 13.9292 ┆ 9.8650 ┆ 35.4335 ┆ 59.5927 ┆ 65.2298 │
│ request inflight       ┆  7547 ┆ 0.0126 ┆  0.3460 ┆  0.0678 ┆  0.0400 ┆ 0.0595 ┆  0.1143 ┆  0.2548 ┆  0.3334 │
│ handle_request         ┆  7548 ┆ 0.0002 ┆  0.0383 ┆  0.0033 ┆  0.0043 ┆ 0.0018 ┆  0.0093 ┆  0.0218 ┆  0.0357 │
│ scan                   ┆    71 ┆ 0.0008 ┆ 13.2204 ┆  3.4702 ┆  3.8108 ┆ 1.5854 ┆  8.5564 ┆ 13.2204 ┆ 13.2204 │
│ merge                  ┆    82 ┆ 0.0009 ┆ 14.7640 ┆  0.3741 ┆  1.9679 ┆ 0.0409 ┆  0.1169 ┆ 14.7640 ┆ 14.7640 │
│ prune                  ┆    74 ┆ 0.0002 ┆  1.3170 ┆  0.1349 ┆  0.2529 ┆ 0.0181 ┆  0.4215 ┆  1.3170 ┆  1.3170 │
│ trash                  ┆    69 ┆ 0.0009 ┆ 10.8716 ┆  2.9331 ┆  3.2820 ┆ 1.1828 ┆  8.1202 ┆ 10.8716 ┆ 10.8716 │
└────────────────────────┴───────┴────────┴─────────┴─────────┴─────────┴────────┴─────────┴─────────┴─────────┘
```

**Comment by inetic** [comment]
Merged the above work in d3515be6
